lib/player.rb

```diff
@@ -12,44 +12,116 @@ def add_ship(ship)
     @ships << ship
   end
 
+
+
   def ship_setup
     if @type == :auto
-      # helper-method: auto_generation CALLS
-      # helper-method: player_place_ships
-      # auto generation
-      p "type auto"
-    elsif @type == :user
-      # Put in a guard for user error
-
-      # helper-method: player_place_ships
-      @ships.each do |ship|
-        if ship.length == 3
-          puts "Enter the squares for the Cruiser (3 spaces):"
-          puts "> "
-        elsif ship.length == 2
-          puts "Enter the squares for the Submarine (2 spaces):"
+      until all_ships_placed?
+        auto_generate_coordinates
+      end
+    else
+      until all_ships_placed?
+        @ships.each do |ship|
+          puts "Enter the squares for the #{ship.name} (#{ship.length} spaces)"
           puts "> "
-        end
-
-        user_ship_coords = gets.chomp.split(/ /)
-        # Consider loop for user error
-        if @board.valid_placement?(ship, user_ship_coords) == true
-          @board.place(ship, user_ship_coords)
-        elsif @board.valid_placement?(ship, user_ship_coords) == false
-          loop do
-            puts "Those are invalid coordinates. Please try again:"
-            puts "> "
-            user_ship_coords = gets.chomp.split(/ /)
-
-            if @board.valid_placement?(ship, user_ship_coords) == true
-              @board.place(ship, user_ship_coords)
-              require "pry"; binding.pry
-              exit # Could be a helper-method
+          user_ship_coords = gets.chomp.upcase.split(/ /)
+          result = @board.valid_placement?(ship, user_ship_coords)
+          if result == false
+            until this_ship_placed?(ship)
+              puts "Those are invalid coordinates. Please try again: "
+              user_ship_coords2 = gets.chomp.upcase.split(/ /)
+            @board.valid_placement?(ship, user_ship_coords2)
+            @board.validated_placement
             end
+          else
+            @board.validated_placement
           end
         end
       end
     end
   end
 
+  def auto_generate_coordinates
```

Comment:
This is amazing! Great job with the auto generation. With your sound logic on this, I cannot wait to attach Computer AI options for iteration 4. Also - solid work here. This is outstanding and feels pretty air tight!
Human speak for my brain:
1. auto_generate_coordinates calls the players ships array and iterates through. For each ship, an until loop is initiated the will break when this_ship_placed?(ship) returns true. 
2. The until loop creates a blank array assigned to auto_coords. 
3. Since @boards.cells.keys will return the array of cell coordinates, possible is now assigned to a randomized array of cell coordinates. 
4. This will now call the ship length, and for it's length integer, it will DO --> a.) shovel in the first element of the randomized coordinates array into the blank auto_coords array. b.) Then using [.rotate!](https://ruby-doc.org/core-2.6.1/Array.html#method-i-rotate-21), transforms the possible array by moving the first element to the end.
5. After the .times do block, the until loop will now check if the auto gen coords placed ships on the player's(in this case :auto) board in the correct place.
6. Now it will place board if valid_placement? returns true (if is built into method).